### DISCUSSION TITLE
Consider extending uBlock Privacy with keyboard-map permission policy

### DISCUSSION THREAD

**Discussion by bershanskiy** [uBlockOrigin/uAssets#20159] [Q&A]
### Prerequisites

- [X] This is not a support issue or a question. For support, questions, or help, visit [/r/uBlockOrigin](https://www.reddit.com/r/uBlockOrigin/).
- [X] I performed a [cursory search of the issue tracker](https://github.com/uBlockOrigin/uAssets/issues?q=is%3Aissue) to avoid opening a duplicate issue.
- [X] I am running the latest version of uBlock Origin (uBO).

### I tried to reproduce the issue when...

- [X] uBO is the only extension.
- [X] uBO uses default lists and settings.
- [X] using a new, unmodified browser profile.

### Description

[ublock-privacy](https://github.com/uBlockOrigin/uAssets/blob/master/filters/privacy.txt) list already has directives for a few privacy-hostile permissions. I propose adding `keyboard-map` directive.

This directive controls Keyboard Map API ([specification](https://wicg.github.io/keyboard-map/)) which specifies `navigator.keyboard.getLayoutMap()` method ([MDN docs](https://developer.mozilla.org/en-US/docs/Web/API/Keyboard/getLayoutMap)). This API exposes a significant amount of fingerprintable data for users of non-standard keyboard layouts or users with unusual keyboards, while not informing the user about its usage. This API has very little practical utility, it can be mostly used just for shortcut management in web productivity programs (think user who uses Dvorak layout pressing Ctrl+Z shortcut but program actually receiving something like Ctrl+Q) and browser video games (like user trying to use AWSD keys for walking but program receiving A,OE keys). These use cases are rater rear/weak and are solved much better via website-specific custom settings UI which offers user to select an action and then press the desired key combination to assign it). Both Mozilla and Apple opposed this feature on these grounds and declined implementing it.

[Mozilla position](https://mozilla.github.io/standards-positions/#keyboard-map):
> We're concerned that this exposes keyboard layouts, which seem likely to be a significant source of fingerprinting data, in a way that does not require any user interaction.

[Apple position](https://github.com/WICG/keyboard-map/issues/30#issue-487691188):
> [T]he WebKit team at Apple is not interested in implementing this feature as currently proposed / spec'ed.

Brave disables Keyboard Map API (sets `navigator.keyboard` to `null`) when their "Brave Shields" are enabled since 2021 ([source](https://github.com/brave/brave-core/pull/10935)).

Example reproduction:
1. Open `https://example.com`
2. Enable a few different keyboards in your OS, e.g. English and Dvorak (DV).
3. Run this in the console and observe the output with different keyboard layouts:
   ```js
   var keyboard = {};
   navigator.keyboard.getLayoutMap()
   .then(keyboardLayoutMap => {
     keyboardLayoutMap.forEach((value,key,map) => {keyboard[key] = value})
     console.log(keyboard)
   });
   ```
4. Observe the following layouts
   For standard English keyboard:
   ```js
   {KeyE: "e", KeyD: "d", Minus: "-", KeyH: "h", KeyZ: "z", ...}
   ```
   For Dvorak keyboard:
   ```
   {KeyE: ".", KeyD: "e", Minus: "[", KeyH: "d", KeyZ: ";", ...}
   ```

### URL(s) where the issue occurs.

`https://example.com`

### uBO version

1.52.2

### Browser name and version

Chromium 118.0.5993.70

### Settings

 - None

**Comment by MasterKia** [comment]
See: https://github.com/WICG/keyboard-map/issues/38

> I'm a developer in the Excel Online application, working specifically on keyboard shortcuts.
>
> Our application is hosted as a sub-context inside of an iframe element in the dome, where the top-level context can be many different endpoints - SharePoint, OneDrive, Teams, etc. That means the API is not available for us to use from inside the app code.
>
> In addition to that, keyCode is deprecated, and there doesn't seem to be any alternative API to handle keyboard shortcuts for different layouts, other than getLayoutMap().
>
> Should the keyCode property actually be removed one day - we believe all Office Online applications (PowerPoint, Word, OneNote, etc.) will require this or similar API to handle their keyboard shortcuts for different layouts.

**Comment by bershanskiy** [comment]
I was playing with this PR and I believe Chromium currently has a bug which prevents Permissions Policy from working. In particular, I see that
1. before this PR, `document.featurePolicy.allowsFeature('keyboard-map')` returns `true` and keyboard map is available, as expected
2. after this PR, `document.featurePolicy.allowsFeature('keyboard-map')` returns `false` but keyboard map is still available, even though the promise should be rejected ([source](https://wicg.github.io/keyboard-map/#keyboard-getlayoutmap))

**Reply by krystian3w** [reply]
How possible check this in Firefox if one day they implement it?

**Reply by gorhill** [reply]
So until they fix this bug, there is not much point to add the filter?

**Comment by bershanskiy** [comment]
> > Should the keyCode property actually be removed one day - we believe all Office Online applications (PowerPoint, Word, OneNote, etc.) will require this or similar API to handle their keyboard shortcuts for different layouts.

Yes, keyboard shortcuts for different layouts is the primary use case for this API. All modern browsers support [KeyboardEvent/code](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code) which is an explicit successor to `KeyboardEvent.keyCode`. Both `KeyboardEvent.code` and `KeyboardEvent.keyCode` expose this information, for better or worse (they can facilitate fingerprinting, but do require user interaction for each key). This is explicitly mentioned in the spec.

**Comment by MasterKia** [comment]
cc @maymen1

**Reply by mapx-** [reply]
@gorhill
